examples/ptb/train_ptb_custom_loop.py

```diff
@@ -0,0 +1,139 @@
+    args = parser.parse_args()
+
+    def evaluate(model, iter):
+        # Evaluation routine to be used for validation and test.
+        model.predictor.train = False
+        evaluator = model.copy()  # to use different state
+        evaluator.predictor.reset_state()  # initialize state
+        evaluator.predictor.train = False  # dropout does nothing
+        sum_perp = 0
+        data_count = 0
+        for batch in copy.copy(iter):
+            x, t = convert.concat_examples(batch, args.gpu)
+            loss = evaluator(x, t)
+            sum_perp += loss.data
+            data_count += 1
+        model.predictor.train = True
+        return sum_perp / data_count
+
+    # Load the Penn Tree Bank long word sequence dataset
+    train, val, test = chainer.datasets.get_ptb_words()
+    n_vocab = max(train) + 1  # train is just an array of integers
+    print('#vocab =', n_vocab)
+
+    if args.test:
+        train = train[:100]
+        val = val[:100]
+        test = test[:100]
+
+    # Create the dataset iterators
+    train_iter = train_ptb.ParallelSequentialIterator(train, args.batchsize)
+    val_iter = train_ptb.ParallelSequentialIterator(val, 1, repeat=False)
+    test_iter = train_ptb.ParallelSequentialIterator(test, 1, repeat=False)
+
+    # Prepare an RNNLM model
+    rnn = train_ptb.RNNForLM(n_vocab, args.unit)
+    model = L.Classifier(rnn)
+    model.compute_accuracy = False  # we only want the perplexity
+    if args.gpu >= 0:
+        chainer.cuda.get_device(args.gpu).use()  # make the GPU current
+        model.to_gpu()
+
+    # Set up an optimizer
+    optimizer = chainer.optimizers.SGD(lr=1.0)
+    optimizer.setup(model)
+    optimizer.add_hook(chainer.optimizer.GradientClipping(args.gradclip))
+
+    sum_perp = 0
+    count = 0
+    iteration = 0
+    while train_iter.epoch < args.epoch:
+        loss = 0
+        iteration += 1
+        # Progress the dataset iterator for bprop_len words at each iteration.
+        for i in range(args.bproplen):
+            # Get the next batch (a list of tuples of two word IDs)
+            batch = train_iter.__next__()
+            # Concatenate the word IDs to matrices and send them to the device
+            # self.converter does this job
+            # (it is chainer.dataset.concat_examples by default)
+            x, t = convert.concat_examples(batch, args.gpu)
+            # Compute the loss at this time step and accumulate it
+            loss += optimizer.target(chainer.Variable(x), chainer.Variable(t))
+            count += 1
+
+        sum_perp += loss.data
+        optimizer.target.cleargrads()  # Clear the parameter gradients
+        loss.backward()  # Backprop
+        loss.unchain_backward()  # Truncate the graph
+        optimizer.update()  # Update the parameters
+
+        if iteration % 10 == 0:
+            print('iteration: ', iteration)
+            print('training perplexity: ', sum_perp / count)
```

Comment:
I fixed the custom loop version to compute the perplexity in the same way as the trainer version.